File: logger/src/main/java/ru/surfstudio/android/logger/logging_strategies/impl/concrete/remote_logger/RemoteLoggerLoggingStrategy.java
package ru.surfstudio.android.logger.logging_strategies.impl.concrete.remote_logger;

import android.util.Log;

import org.jetbrains.annotations.Nullable;

import ru.surfstudio.android.logger.RemoteLogger;
import ru.surfstudio.android.logger.logging_strategies.impl.base.AbstractGroupedLoggingStrategy;

public class RemoteLoggerLoggingStrategy extends AbstractGroupedLoggingStrategy {

    private static final String REMOTE_LOGGER_LOG_FORMAT = "%s: %s";
    private static final int MINIMAL_LOG_PRIORITY = Log.DEBUG;

    private final ThreadLocal<String> explicitTag = new ThreadLocal<>();

    @Override
    protected void log(int priority, Throwable t) {
        log(priority, t, null);
    }

    @Override
    protected void log(int priority, String message, Object... args) {
        log(priority, null, message, args);
    }

    @Override
    protected void log(int priority, Throwable t, String message, Object... args) {
        if (priority < MINIMAL_LOG_PRIORITY) {
            return;
        }

        if (message != null) {
            RemoteLogger.logMessage(String.format(REMOTE_LOGGER_LOG_FORMAT, getTag(), message));
        }

        if (t != null && priority >= Log.ERROR) {
            RemoteLogger.logError(t);
        }
    }

    @Nullable
    private String getTag() {
        String tag = explicitTag.get();
        if (tag != null) {
            explicitTag.remove();
        }
        return tag;
    }
}

File: template/base_feature/src/main/java/ru/surfstudio/standard/application/logger/strategies/remote/RemoteLoggerLoggingStrategy.kt
package ru.surfstudio.standard.application.logger.strategies.remote

import android.util.Log

import ru.surfstudio.standard.base.logger.RemoteLogger
import ru.surfstudio.android.logger.logging_strategies.LoggingStrategy

/**
 * Logging strategy for [RemoteLogger] usage for messages with priority since Log.DEBUG
 */
class RemoteLoggerLoggingStrategy : LoggingStrategy {

    private val explicitTag = ThreadLocal<String>()

    private val tag: String?
        get() {
            val tag = explicitTag.get()
            if (tag != null) {
                explicitTag.remove()
            }
            return tag
        }

    override fun log(priority: Int, t: Throwable?, message: String?, vararg args: Any?) {
        if (priority < MINIMAL_LOG_PRIORITY) {
            return
        }

        if (message != null) {
            RemoteLogger.logMessage(String.format(REMOTE_LOGGER_LOG_FORMAT, tag, message))
        }

        if (t != null && priority >= Log.ERROR) {
            RemoteLogger.logError(t)
        }
    }

    companion object {

        private val REMOTE_LOGGER_LOG_FORMAT = "%s: %s"
        private val MINIMAL_LOG_PRIORITY = Log.DEBUG
    }
}
